### DISCUSSION TITLE
strongswan keys creation

### DISCUSSION THREAD

**Discussion by musaleh17** [strongswan/strongswan#1484] [Q&A]
Hi,

I was wondering if strongswan library could be used for keys exchange and creating AES keys only?

**Comment by tobiasbrunner** [comment]
In what context?

**Reply by musaleh17** [reply]
I have an application that does the encryption part for sessions, I need a daemon for key exchange, and I am wondering if I can use strongswan for creating the shared key between two sides.

**Reply by tobiasbrunner** [reply]
I guess so.

**Reply by musaleh17** [reply]
do you have any example of such a similar case? 
or if it is documented somewhere how it could be added?

**Reply by tobiasbrunner** [reply]
Not really, but strongSwan obviously does this itself (the IKE daemons/clients link the libraries, which use the crypto plugins). But it will depend on a lot of things (e.g. on which layer you actually want to implement this, just the key exchange or actually use IKEv2). See [here](https://docs.strongswan.org/docs/5.9/devs/devs.html) for some developer docs.

**Reply by musaleh17** [reply]
thanks, additional question, is there a way to know what is the IPsec SA for a specific session through the strongswan tools?

**Reply by tobiasbrunner** [reply]
What do you mean?

**Reply by musaleh17** [reply]
According to the Stronswan doc "strongSwan installs the negotiated IPsec SAs and SPs into the kernel", I am checking if I can replace the kernel with my IPsec app? which means to strongswan for creating the IPSEC SA and to inject it to my app instead of the kernel. 
is it something possible? is there a way to integrate my app with strongswan?

**Reply by tobiasbrunner** [reply]
You could write a custom implementation of the `kernel_ipsec_t` interface (there are several examples in our code base).
